jiant/tasks/lib/templates/edge_probing.py

```diff
@@ -0,0 +1,154 @@
+"""Two-span, multi-label edge-probing task base class.
+
+This module defines an AbstractProbingTask and inheritable logic for tokenization and featurization.
+Two-span, multi-label edge-probing tasks should inherit from AbstractProbingTask and the dataclasses
+defined here. See spr1.py for an example task implementation inheriting from this module.
+
+"""
+from abc import ABC
+
+import numpy as np
+import torch
+from dataclasses import dataclass
+from typing import List, Tuple
+
+from jiant.tasks.core import (
+    BaseExample,
+    BaseTokenizedExample,
+    BaseDataRow,
+    BatchMixin,
+    Task,
+    TaskTypes,
+)
+from jiant.tasks.lib.templates.shared import (
+    create_input_set_from_tokens_and_segments,
+    add_cls_token,
+)
+from jiant.tasks.utils import ExclusiveSpan, truncate_sequences
+from jiant.utils import retokenize
+
+
+@dataclass
+class Example(BaseExample):
+    guid: str
+    text: str
+    span1: List[int]
+    span2: List[int]
+    labels: List[str]
+
+    @property
+    def task(self):
+        raise NotImplementedError()
+
+    def tokenize(self, tokenizer):
+        native_tokenization = self.text.split()
+        target_tokenization = tokenizer.tokenize(self.text)
+        aligner = retokenize.TokenAligner(native_tokenization, target_tokenization)
+        target_span1 = aligner.project_span(self.span1[0], self.span1[1])
+        target_span2 = aligner.project_span(self.span2[0], self.span2[1])
+        return TokenizedExample(
+            guid=self.guid,
+            tokens=target_tokenization,
+            span1_span=target_span1,
+            span2_span=target_span2,
+            span1_text=" ".join(target_tokenization[target_span1[0] : target_span1[1]]),
+            span2_text=" ".join(target_tokenization[target_span2[0] : target_span2[1]]),
+            label_ids=[self.task.LABEL_TO_ID[label] for label in self.labels],
+            label_num=len(self.task.LABELS),
```

Comment:
@zphang — is there a better way for me to get access to the number of labels for use in the featurization stage (than passing them through here)?